server/oauth2_test.go

```diff
@@ -263,11 +264,70 @@ func TestValidRedirectURI(t *testing.T) {
 			redirectURI: "http://foo.com/bar",
 			wantValid:   true,
 		},
+		{
+			// check https is valid protocol
+			client: storage.Client{
+				RedirectURIs: []string{"https://foo.com/bar/baz"},
+			},
+			redirectURI: "https://foo.com/bar/baz",
+			wantValid:   true,
+		},
 		{
 			client: storage.Client{
 				RedirectURIs: []string{"http://foo.com/bar"},
 			},
 			redirectURI: "http://foo.com/bar/baz",
+			wantValid:   false,
+		},
+		{
+			client: storage.Client{
+				RedirectURIs: []string{"http://*.foo.com/bar"},
+			},
+			redirectURI: "http://abc.foo.com/bar",
+			wantValid:   true,
+		},
+		{
+			// URL must use http: or https: protocol
+			client: storage.Client{
+				RedirectURIs: []string{"unknown://*.foo.com/bar"},
+			},
+			redirectURI: "unknown://abc.foo.com/bar",
+			wantValid:   false,
+		},
+		{
+			// wildcard must be located in the subdomain furthest from the root domain.
+			//  https://abc.*.foo.com not ok.
+			client: storage.Client{
+				RedirectURIs: []string{"http://abc.*.foo.com/bar"},
+			},
+			redirectURI: "http://abc.123.foo.com/bar",
+			wantValid:   false,
+		},
+		{
+			// wildcard may be prefixed and/or suffixed with additional valid hostname characters
+			//  https://pre-*-post.foo.com will work
+			client: storage.Client{
+				RedirectURIs: []string{"http://pre-*-post.foo.com/bar"},
+			},
+			redirectURI: "http://pre-dinner-post.foo.com/bar",
+			wantValid:   true,
+		},
+		{
+			// valid wildcard will not match a URL more than one subdomain level in place of the wildcard
+			//  https://*.foo.com will not work with https://123.abc.foo.com.
+			client: storage.Client{
+				RedirectURIs: []string{"https://*.foo.com/"},
```

Comment:
In our environment, we are currently authenticating second level hosts via OIDC using wildcards, eg, "https://hostname.namespace.k8s-cluster-name/".

Could we have an option to allow this by making it explicit that we want to match dots as well? maybe something like "https://**.foo.com/"?

Reply:
Yes, this seems reasonable to me.  I'll update test cases & docs and last modify core flow.

Reply:
I think explicit wildcards would be better, i.e.
``
https://*.*.foo.com/
```

Then the prefix and suffix function will still work / be available.

Reply:
To limit PR scope, I think it might be better to file new issue to consider multiple subdomains and modify as separate ticket as being an additional feature?

The change will be an enhancement and backwards comparible, so two steps seems logical?

Reply:
I support the idea of `*.*.foo.com`, however, I can see that causing a set of rules like *.foo.com, *.*.foo.com, *.*.*.foo.com, *.*.*.*.foo.com, depending on how people are using their domain names.

The **.foo.com syntax was loosely based off of shell globbing syntax (**/*.py).

I'm on vacation right now but when I get home I'll write more (maybe even some code! lol).

Currently, we've forked Dex so that we can support allowing it to authenticate entire subdomain trees.

Reply:
I've had a look further into this, and I have some concerns about the performance of the shell globbing syntax using the current approach, because of the runtime overhead associated with it.  Effectively a fork on the standard Go URL.parse function needs to be written in terms of carving out the scheme (as `*` is not permitted in hostname and returns nil).

https://tools.ietf.org/html/rfc3986#section-3.3

I'll have a quick look at whether it can be done efficiently today.

Reply:
Both are implemented; better caching; better comparison (use host portion only of URI) - conservatively implemented.
It would be good to have access to the logger somehow for debug statements, but could not see how this should be done cleanly.  Any input welcome!